File: java/merge_sort_step.java
import java.io.*;
import java.util.*;

public class merge_sort_step {
    public static class Pair {
        int i;
        String s;

        public Pair(int i, String s) {
            this.i = i;
            this.s = s;
        }

        @Override
        public String toString() {
            return i + "," + s;
        }

        public String toFormatted() {
            return i + "/" + s;
        }
    }

    public static void main(String[] args) {
        final String inputFile = "dataset_sample_1000.csv";
        final int startRow = 1; // inclusive, 1-based
        final int endRow = 7;   // inclusive
        final String outputFile = String.format("merge_sort_step_%d_%d.txt", startRow, endRow);

        ArrayList<Pair> data = readCSV(inputFile, startRow, endRow);

        List<String> stepLogs = new ArrayList<>();
        stepLogs.add(toStepString(data)); // initial state

        mergeSort(data, 0, data.size() - 1, stepLogs);

        writeSteps(stepLogs, outputFile);
    }

    // 读取指定范围的数据行
    public static ArrayList<Pair> readCSV(String filename, int start, int end) {
        ArrayList<Pair> list = new ArrayList<>();
        try (BufferedReader br = new BufferedReader(new FileReader(filename))) {
            String line;
            int currentRow = 1;
            while ((line = br.readLine()) != null) {
                if (currentRow >= start && currentRow <= end) {
                    String[] parts = line.split(",", 2);
                    if (parts.length == 2) {
                        int number = Integer.parseInt(parts[0].trim());
                        String word = parts[1].trim();
                        list.add(new Pair(number, word));
                    }
                }
                currentRow++;
            }
        } catch (IOException e) {
            System.out.println("Error reading file: " + e.getMessage());
        }
        return list;
    }

    // Merge Sort 过程
    public static void mergeSort(ArrayList<Pair> arr, int left, int right, List<String> logs) {
        if (left < right) {
            int mid = (left + right) / 2;
            mergeSort(arr, left, mid, logs);
            mergeSort(arr, mid + 1, right, logs);
            merge(arr, left, mid, right, logs);
        }
    }

    // 合并两个子数组并记录当前排序状态
    public static void merge(ArrayList<Pair> arr, int left, int mid, int right, List<String> logs) {
        List<Pair> leftList = new ArrayList<>(arr.subList(left, mid + 1));
        List<Pair> rightList = new ArrayList<>(arr.subList(mid + 1, right + 1));

        int i = 0, j = 0, k = left;
        while (i < leftList.size() && j < rightList.size()) {
            if (leftList.get(i).i <= rightList.get(j).i) {
                arr.set(k++, leftList.get(i++));
            } else {
                arr.set(k++, rightList.get(j++));
            }
        }
        while (i < leftList.size()) arr.set(k++, leftList.get(i++));
        while (j < rightList.size()) arr.set(k++, rightList.get(j++));

        logs.add(toStepString(arr));
    }

    // 将当前数组状态转为格式化字符串
    public static String toStepString(List<Pair> list) {
        List<String> formatted = new ArrayList<>();
        for (Pair p : list) {
            formatted.add(p.toFormatted());
        }
        return String.join(", ", formatted);
    }

    // 写入所有排序步骤到文件
    public static void writeSteps(List<String> steps, String filename) {
        try (BufferedWriter bw = new BufferedWriter(new FileWriter(filename))) {
            for (String step : steps) {
                bw.write(step);
                bw.newLine();
            }
        } catch (IOException e) {
            System.out.println("Error writing file: " + e.getMessage());
        }
    }
}

File: cppmergesort/merge_sort_step.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <string>
#include <sstream>

using namespace std;

struct Data {
    int number;
    string text;
};

vector<string> stepList;

void recordStep(const vector<Data>& arr) {
    string line;
    for (int i = 0; i < arr.size(); ++i) {
        line += to_string(arr[i].number) + "/" + arr[i].text;
        if (i != arr.size() - 1) line += ", ";
    }
    stepList.push_back(line);
}

void merge(vector<Data>& arr, int l, int m, int r) {
    int len1 = m - l + 1;
    int len2 = r - m;
    vector<Data> L(len1), R(len2);

    for (int i = 0; i < len1; i++) L[i] = arr[l + i];
    for (int i = 0; i < len2; i++) R[i] = arr[m + 1 + i];

    int i = 0, j = 0, k = l;
    while (i < len1 && j < len2) {
        if (L[i].number <= R[j].number) {
            arr[k++] = L[i++];
        } else {
            arr[k++] = R[j++];
        }
    }

    while (i < len1) arr[k++] = L[i++];
    while (j < len2) arr[k++] = R[j++];

    recordStep(arr); // after merge
}

void mergeSort(vector<Data>& arr, int l, int r) {
    if (l < r) {
        int mid = (l + r) / 2;
        mergeSort(arr, l, mid);
        mergeSort(arr, mid + 1, r);
        merge(arr, l, mid, r);
    }
}

vector<Data> readSelectedRows(string fileName, int from, int to) {
    ifstream file(fileName);
    vector<Data> list;
    string line;
    int count = 0;

    while (getline(file, line)) {
        if (count + 1 >= from && count + 1 <= to) {
            stringstream ss(line);
            string numStr, txt;
            getline(ss, numStr, ',');
            getline(ss, txt);
            Data row;
            row.number = stoi(numStr);
            row.text = txt;
            list.push_back(row);
        }
        count++;
        if (count >= to) break;
    }

    file.close();
    return list;
}

int main(int argc, char* argv[]) {
    if (argc != 4) {
        cout << "Usage: ./merge_sort_step <input_file> <start_row> <end_row>" << endl;
        return 1;
    }

    string inputFile = argv[1];
    int startRow = stoi(argv[2]);
    int endRow = stoi(argv[3]);

    vector<Data> records = readSelectedRows(inputFile, startRow, endRow);

    recordStep(records); // log the initial state
    mergeSort(records, 0, records.size() - 1);

    ofstream out("merge_step_output.txt");
    for (int i = 0; i < stepList.size(); i++) {
        out << stepList[i] << endl;
    }
    out.close();

    cout << "Steps written to merge_step_output.txt" << endl;
    return 0;
}


// g++ merge_sort_step.cpp -o merge_sort_step
// ./merge_sort_step dataset_sample_1000.csv 1 7